File: C++/031. Next Permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int size = nums.size();
        if(size<=1){return;}
        int index1=size-1;
        while(index1>0)
        {
            if(nums[index1]>nums[index1-1]){break;}
            index1--;
        }
        if(index1==0)
        {
            sort(nums.begin(),nums.end());
            return;
        }
        int index2=size-1;
        while(index2>=index1)
        {
            if (nums[index2]>nums[index1-1]){break;}
            index2--;
        }
        swap(nums[index1-1],nums[index2]);
        sort(nums.begin()+index1,nums.end());
        return;
    }
};

File: python/031. Next Permutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        idx1 = n-1
        # find the first (backwards) idx1 such that nums[idx1] > nums[idx1-1]
        while idx1 > 0 and nums[idx1] <= nums[idx1-1]:
            idx1 -= 1
        if idx1 > 0:
            # find the first (backwards) idx2 such that nums[idx2] > nums[idx1-1]
            idx2 = n-1
            while idx2 > idx1 and nums[idx2] <= nums[idx1-1]:
                idx2 -= 1
            nums[idx1-1], nums[idx2] = nums[idx2], nums[idx1-1]
        # sort nums[idx1:]
        for i in range((n-idx1)/2):
            nums[idx1+i], nums[n-1-i] = nums[n-1-i], nums[idx1+i]